File: 백준/Bronze/11948. 과목선택/과목선택.go
package main

import (
	"bufio"
	"cmp"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Problem struct {
	*bufio.Reader
	*bufio.Writer
	*StringBuilder
	*Data
}

type Data struct {
	sci []int
	his []int
	sum int
}

func NewProblem() *Problem {
	return &Problem{
		Reader:        bufio.NewReader(os.Stdin),
		Writer:        bufio.NewWriter(os.Stdout),
		StringBuilder: NewStringBuilder(),
		Data:          NewData(),
	}
}

func NewData() *Data {
	return &Data{}
}

func (p *Problem) Solve() {
	p.input()
	p.logic()
	p.output()
}

func (p *Problem) input() {
	var v int
	for i := 0; i < 4; i++ {
		if _, err := fmt.Fscan(p.Reader, &v); err != nil {
			os.Exit(1)
		}
		p.sci = append(p.sci, v)
		p.sum += v
	}
	for i := 0; i < 2; i++ {
		if _, err := fmt.Fscan(p.Reader, &v); err != nil {
			os.Exit(1)
		}
		p.his = append(p.his, v)
		p.sum += v
	}
}

func (p *Problem) logic() {
	sci := Min(p.sci[0], p.sci[1:]...)
	his := Min(p.his[0], p.his[1:]...)
	p.Append(p.sum - sci - his)
}

func (p *Problem) output() {
	_, err := fmt.Fprint(p.Writer, p.String())
	if err != nil {
		os.Exit(1)
	}
}

func main() {
	p := NewProblem()
	defer func(p *Problem) {
		_ = p.Flush()
	}(p)
	p.Solve()
}

func Min[T cmp.Ordered](v1 T, v2 ...T) T {
	for _, v := range v2 {
		if cmp.Compare(v1, v) > 0 {
			v1 = v
		}
	}
	return v1
}

func Max[T cmp.Ordered](v1 T, v2 ...T) T {
	for _, v := range v2 {
		if cmp.Compare(v, v1) > 0 {
			v1 = v
		}
	}
	return v1
}

type StringBuilder struct {
	strings.Builder
}

func NewStringBuilder() *StringBuilder {
	var sb StringBuilder
	sb.Grow(30)
	return &sb
}

func (sb *StringBuilder) Append(values ...interface{}) *StringBuilder {
	for _, value := range values {
		sb.WriteString(convert(value))
	}
	return sb
}

func (sb *StringBuilder) ReverseAppend(values ...interface{}) *StringBuilder {
	for _, value := range values {
		str := convert(value)
		var res string
		for _, s := range str {
			res = string(s) + res
		}
		sb.WriteString(res)
	}
	return sb
}

func convert(value interface{}) string {
	switch v := value.(type) {
	case string:
		return v
	case int:
		return strconv.Itoa(v)
	case int8:
		return strconv.Itoa(int(v))
	case int16:
		return strconv.Itoa(int(v))
	case int32:
		return strconv.Itoa(int(v))
	case int64:
		return strconv.FormatInt(v, 10)
	case uint:
		return strconv.FormatUint(uint64(v), 10)
	case uint8:
		return strconv.FormatUint(uint64(v), 10)
	case uint16:
		return strconv.FormatUint(uint64(v), 10)
	case uint32:
		return strconv.FormatUint(uint64(v), 10)
	case uint64:
		return strconv.FormatUint(v, 10)
	case float32:
		return strconv.FormatFloat(float64(v), 'f', -1, 32)
	case float64:
		return strconv.FormatFloat(v, 'f', -1, 64)
	default:
		return fmt.Sprintf("%v", v)
	}
}

File: 백준/Bronze/11948. 과목선택/과목선택.rs
use std::io;
use std::io::{BufReader, BufWriter, Read, StdinLock, StdoutLock, Write};

struct Problem<'a> {
    reader: BufReader<StdinLock<'a>>,
    writer: BufWriter<StdoutLock<'a>>,
    data: Vec<u32>,
    answer: String,
}

impl<'a> Problem<'a> {
    fn new() -> Problem<'a> {
        Problem {
            reader: BufReader::new(io::stdin().lock()),
            writer: BufWriter::new(io::stdout().lock()),
            data: Vec::new(),
            answer: String::new(),
        }
    }

    fn input(&mut self) {
        let mut input = String::new();
        self.reader.read_to_string(&mut input).unwrap();
        let mut input = input
            .split_ascii_whitespace()
            .map(|e| e.parse::<u32>().unwrap());
        let mut input = || input.next().unwrap();
        for _i in 0..6 {
            self.data.push(input());
        }
    }

    fn solve(&mut self) {
        let sci = self.min(0, 4);
        let his = self.min(4, 6);
        let sum: u32 = self.data.iter().sum();
        self.answer.push_str(&*(sum - sci - his).to_string());
    }

    fn min(&self, begin: u32, end: u32) -> u32 {
        let mut min = self.data[begin as usize];
        for i in begin + 1..end {
            if min > self.data[i as usize] {
                min = self.data[i as usize];
            }
        }
        min
    }

    fn output(&mut self) {
        writeln!(self.writer, "{}", self.answer).unwrap();
    }
}

fn main() {
    let mut problem = Problem::new();
    problem.input();
    problem.solve();
    problem.output();
}
